src/worker.cc

```diff
@@ -0,0 +1,715 @@
+}
+
+void WorkerContext::Terminate(const FunctionCallbackInfo<Value>& args) {
+  WorkerContext* context = node::Unwrap<WorkerContext>(args.Holder());
+  if (context == nullptr)
+    return;
+  context->Terminate();
+}
+
+void WorkerContext::Terminate(bool should_emit_exit_event) {
+  CHECK_CALLED_FROM_OWNER(this);
+  ScopedLock::Mutex init_lock(initialization_mutex());
+  ScopedLock::Mutex api_lock(api_mutex());
+  ScopedLock::Mutex term_lock(termination_mutex());
+
+  if (termination_kind() != NONE)
+    return;
+
+  set_termination_kind(ABRUPT);
+
+  if (is_initialized()) {
+    should_emit_exit_event_ = should_emit_exit_event;
+    worker_isolate()->TerminateExecution();
+    worker_notifications()->Notify();
+  }
+}
+
+bool WorkerContext::ProcessMessageToWorker(Isolate* isolate,
+                                           WorkerMessage* message) {
+  Local<Value> argv[] = {
+    String::NewFromUtf8(isolate, message->payload()),
+    Integer::New(isolate, message->type())
+  };
+  delete message;
+  MakeCallback(isolate,
+               worker_wrapper(),
+               "_onmessage",
+               ARRAY_SIZE(argv),
+               argv);
+  return worker_env()->CanCallIntoJs();
+}
+
+void WorkerContext::ProcessMessagesToWorker() {
+  CHECK_CALLED_FROM_WORKER(this);
+
+  Isolate* isolate = worker_env()->isolate();
+  HandleScope scope(isolate);
+  Context::Scope context_scope(worker_env()->context());
+
+  flush_primary_queue:
+  uint32_t messages_processed = 0;
+  while (WorkerMessage* message = to_worker_messages_primary_.PopFront()) {
+    if (!ProcessMessageToWorker(isolate, message))
+      return;
+    messages_processed++;
+  }
+
+  if (messages_processed >= kMaxPrimaryQueueMessages) {
+    while (true) {
+      WorkerMessage* message = nullptr;
+      {
+        ScopedLock::Mutex lock(to_worker_messages_mutex());
+        message = to_worker_messages_backup_.PopFront();
+      }
+
+      if (message == nullptr)
+        goto flush_primary_queue;
+
+      if (!ProcessMessageToWorker(isolate, message))
+        return;
+    }
+  }
+}
+
+bool WorkerContext::ProcessMessageToOwner(Isolate* isolate,
+                                          WorkerMessage* message) {
+  Local<Value> argv[] = {
+    String::NewFromUtf8(isolate, message->payload()),
+    Integer::New(isolate, message->type())
+  };
+  delete message;
+  MakeCallback(isolate,
+               owner_wrapper(),
+               "_onmessage",
+               ARRAY_SIZE(argv),
+               argv);
+  return owner_env()->CanCallIntoJs();
+}
+
+void WorkerContext::ProcessMessagesToOwner() {
+  CHECK_CALLED_FROM_OWNER(this);
+  Isolate* isolate = owner_env()->isolate();
+  HandleScope scope(isolate);
+  Context::Scope context_scope(owner_env()->context());
+
+  flush_primary_queue:
+  uint32_t messages_processed = 0;
+  while (WorkerMessage* message = to_owner_messages_primary_.PopFront()) {
+    if (!ProcessMessageToOwner(isolate, message))
+      return;
```

Comment:
I am not sure what is implied by your second paragraph? After secondary queue is processed, primary queue processing starts again from the `goto`. There should not be any messages left unprocsessed if that's what you are implying.